File: src/posix-extras-accounts/libservice.cpp
#define WORKAROUND_INNER_CLASS_DEFINITION_Accounts__Service
#include <QDomDocument>
#include <QSet>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <service.h>
#include "libservice.h"
#include "libservice.hxx"

Accounts__Service* Accounts__Service_new() {
    return new Accounts::Service();
}

Accounts__Service* Accounts__Service_new2(const Accounts__Service* other) {
    return new Accounts::Service(*other);
}

void Accounts__Service_OperatorAssign(Accounts__Service* self, const Accounts__Service* other) {
    self->operator=(*other);
}

bool Accounts__Service_IsValid(const Accounts__Service* self) {
    return self->isValid();
}

libqt_string Accounts__Service_Name(const Accounts__Service* self) {
    QString _ret = self->name();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string Accounts__Service_Description(const Accounts__Service* self) {
    QString _ret = self->description();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string Accounts__Service_DisplayName(const Accounts__Service* self) {
    QString _ret = self->displayName();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string Accounts__Service_TrCatalog(const Accounts__Service* self) {
    QString _ret = self->trCatalog();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string Accounts__Service_ServiceType(const Accounts__Service* self) {
    QString _ret = self->serviceType();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string Accounts__Service_Provider(const Accounts__Service* self) {
    QString _ret = self->provider();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string Accounts__Service_IconName(const Accounts__Service* self) {
    QString _ret = self->iconName();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

bool Accounts__Service_HasTag(const Accounts__Service* self, const libqt_string tag) {
    QString tag_QString = QString::fromUtf8(tag.data, tag.len);
    return self->hasTag(tag_QString);
}

libqt_list /* set of libqt_string */ Accounts__Service_Tags(const Accounts__Service* self) {
    QSet<QString> _ret = self->tags();
    // Convert QSet<> from C++ memory to manually-managed C memory
    libqt_string* _arr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * _ret.size()));
    int _ctr = 0;
    QSetIterator<QString> _itr(_ret);
    while (_itr.hasNext()) {
        QString _sv_ret = _itr.next();
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _sv_b = _sv_ret.toUtf8();
        libqt_string _sv_str;
        _sv_str.len = _sv_b.length();
        _sv_str.data = static_cast<const char*>(malloc(_sv_str.len + 1));
        memcpy((void*)_sv_str.data, _sv_b.data(), _sv_str.len);
        ((char*)_sv_str.data)[_sv_str.len] = '\0';
        _arr[_ctr++] = _sv_str;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

QDomDocument* Accounts__Service_DomDocument(const Accounts__Service* self) {
    return new QDomDocument(self->domDocument());
}

void Accounts__Service_Delete(Accounts__Service* self) {
    delete self;
}

File: src/posix-extras-accounts/libservice.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");
const set_constu8 = std.StringHashMapUnmanaged(void);

/// https://accounts-sso.gitlab.io/libaccounts-qt/classAccounts_1_1Service.html
pub const accounts__service = struct {
    /// New constructs a new Accounts::Service object.
    ///
    ///
    pub fn New() QtC.Accounts__Service {
        return qtc.Accounts__Service_new();
    }

    /// New2 constructs a new Accounts::Service object.
    ///
    /// ``` other: QtC.Accounts__Service ```
    pub fn New2(other: ?*anyopaque) QtC.Accounts__Service {
        return qtc.Accounts__Service_new2(@ptrCast(other));
    }

    /// [Upstream resources](https://accounts-sso.gitlab.io/libaccounts-qt/classAccounts_1_1Service.html)
    ///
    /// ``` self: QtC.Accounts__Service, other: QtC.Accounts__Service ```
    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.Accounts__Service_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Upstream resources](https://accounts-sso.gitlab.io/libaccounts-qt/classAccounts_1_1Service.html)
    ///
    /// ``` self: QtC.Accounts__Service ```
    pub fn IsValid(self: ?*anyopaque) bool {
        return qtc.Accounts__Service_IsValid(@ptrCast(self));
    }

    /// [Upstream resources](https://accounts-sso.gitlab.io/libaccounts-qt/classAccounts_1_1Service.html)
    ///
    /// ``` self: QtC.Accounts__Service, allocator: std.mem.Allocator ```
    pub fn Name(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Accounts__Service_Name(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("accounts::service.Name: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Upstream resources](https://accounts-sso.gitlab.io/libaccounts-qt/classAccounts_1_1Service.html)
    ///
    /// ``` self: QtC.Accounts__Service, allocator: std.mem.Allocator ```
    pub fn Description(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Accounts__Service_Description(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("accounts::service.Description: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Upstream resources](https://accounts-sso.gitlab.io/libaccounts-qt/classAccounts_1_1Service.html)
    ///
    /// ``` self: QtC.Accounts__Service, allocator: std.mem.Allocator ```
    pub fn DisplayName(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Accounts__Service_DisplayName(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("accounts::service.DisplayName: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Upstream resources](https://accounts-sso.gitlab.io/libaccounts-qt/classAccounts_1_1Service.html)
    ///
    /// ``` self: QtC.Accounts__Service, allocator: std.mem.Allocator ```
    pub fn TrCatalog(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Accounts__Service_TrCatalog(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("accounts::service.TrCatalog: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Upstream resources](https://accounts-sso.gitlab.io/libaccounts-qt/classAccounts_1_1Service.html)
    ///
    /// ``` self: QtC.Accounts__Service, allocator: std.mem.Allocator ```
    pub fn ServiceType(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Accounts__Service_ServiceType(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("accounts::service.ServiceType: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Upstream resources](https://accounts-sso.gitlab.io/libaccounts-qt/classAccounts_1_1Service.html)
    ///
    /// ``` self: QtC.Accounts__Service, allocator: std.mem.Allocator ```
    pub fn Provider(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Accounts__Service_Provider(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("accounts::service.Provider: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Upstream resources](https://accounts-sso.gitlab.io/libaccounts-qt/classAccounts_1_1Service.html)
    ///
    /// ``` self: QtC.Accounts__Service, allocator: std.mem.Allocator ```
    pub fn IconName(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Accounts__Service_IconName(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("accounts::service.IconName: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Upstream resources](https://accounts-sso.gitlab.io/libaccounts-qt/classAccounts_1_1Service.html)
    ///
    /// ``` self: QtC.Accounts__Service, tag: []const u8 ```
    pub fn HasTag(self: ?*anyopaque, tag: []const u8) bool {
        const tag_str = qtc.libqt_string{
            .len = tag.len,
            .data = tag.ptr,
        };
        return qtc.Accounts__Service_HasTag(@ptrCast(self), tag_str);
    }

    /// [Upstream resources](https://accounts-sso.gitlab.io/libaccounts-qt/classAccounts_1_1Service.html)
    ///
    /// ``` self: QtC.Accounts__Service, allocator: std.mem.Allocator ```
    pub fn Tags(self: ?*anyopaque, allocator: std.mem.Allocator) set_constu8 {
        const _set: qtc.libqt_list = qtc.Accounts__Service_Tags(@ptrCast(self));
        var _ret: set_constu8 = .empty;
        const _data: [*]qtc.libqt_string = @ptrCast(@alignCast(_set.data));
        for (0.._set.len) |i| {
            _ret.put(allocator, _data[i].data[0.._data[i].len], {}) catch @panic("accounts::service.Tags: Set insertion failed");
        }
        return _ret;
    }

    /// [Upstream resources](https://accounts-sso.gitlab.io/libaccounts-qt/classAccounts_1_1Service.html)
    ///
    /// ``` self: QtC.Accounts__Service ```
    pub fn DomDocument(self: ?*anyopaque) QtC.QDomDocument {
        return qtc.Accounts__Service_DomDocument(@ptrCast(self));
    }

    /// [Upstream resources](https://accounts-sso.gitlab.io/libaccounts-qt/classAccounts_1_1Service.html)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.Accounts__Service ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.Accounts__Service_Delete(@ptrCast(self));
    }
};
